lms/events/__init__.py

```diff
@@ -1,70 +1,77 @@
-from dataclasses import dataclass
+from dataclasses import asdict, dataclass, field
+from typing import List, Optional
 
 from pyramid.events import subscriber
 from pyramid.request import Request
 
-from lms.models import Event, EventType, LtiLaunches
+from lms.models import EventType, LtiLaunches
 from lms.services import EventService, LTIRoleService
 
 
 @dataclass
-class LTILaunchEvent:
+class BaseEvent:  # pylint:disable=too-many-instance-attributes
     request: Request
+    type: EventType.Type
+
+    user_id: Optional[int] = None
+    role_ids: List[int] = field(default_factory=list)
+    application_instance_id: Optional[int] = None
+    course_id: Optional[int] = None
+    assignment_id: Optional[int] = None
+    grouping_id: Optional[int] = None
+    data: Optional[dict] = None
 
 
 @dataclass
-class LTIDeepLinkingEvent:
-    request: Request
-    document_url: str
+class LTIEvent(BaseEvent):
+    def __post_init__(self):
+        # pylint:disable=no-member
+        if not self.user_id:
+            self.user_id = self.request.user.id
+
+        if not self.role_ids:
```

Comment:
This nesting is not very satisfying too look at but it's not complicated it does the trick.